### PR TITLE
global: addition of ``references`` endpoint for literature

### PR DESCRIPTION
**Pull request by drjova**
* Addition of ``references`` endpoint for literature.

Signed-off-by: Harris Tzovanakis <me@drjova.com>

### CODE REVIEW

**Comment by drjova** [comment]
ci test please

File: inspirehep/modules/records/serializers/schemas/json.py [hunk 1]
```diff
@@ -36,3 +41,98 @@ class RecordSchemaJSONUIV1(Schema):
...
+                    reference_record_id = get_recid_from_ref(
+                        reference.get('record'))
+                    reference_record = recid_to_reference.get(
+                        reference_record_id)
+
+                    if reference_record:
+                        resolved.append(reference_record)
+                        break
+
+                if 'reference' in reference:
```
Comment (michamos) [inline inspirehep/modules/records/serializers/schemas/json.py]: I think this would be clearer if you did this part in a more declarative style. Currently it's quite hard to follow.
Something like
```python
publication_info = force_list(
    reference['publication_info'] if {'journal_title', 'pubinfo_freetext'}.issubset(reference.get('publication_info', {}).keys()) else None
)
# etc.
reference.update({
    'publication_info': publication_info,
    # etc.
})
del reference['reference']
strip_empty_values(reference)
```

File: inspirehep/modules/records/serializers/schemas/json.py [hunk 2]
```diff
@@ -36,3 +41,98 @@ class RecordSchemaJSONUIV1(Schema):
...
+                    reference.update(reference['reference'])
+                    del reference['reference']
+
+                if 'publication_info' in reference:
+                    publication_info = reference['publication_info']
+
+                    if 'journal_title' in publication_info or \
+                       'pubinfo_freetext' in publication_info:
```
Comment (michamos) [inline inspirehep/modules/records/serializers/schemas/json.py]: what about the case of conferences, where you have a `conference_record` but no `journal_title`?

File: inspirehep/modules/theme/views.py [hunk 3]
```diff
@@ -350,7 +350,10 @@ def ping():
 
 @blueprint.route('/ajax/references', methods=['GET'])
 def ajax_references():
-    """Handler for datatables references view"""
+    """Handler for datatables references view.
+
+    DEPRECATED: REMOVE WHEN THE UI IS RELEASED.
```
Comment (michamos) [inline inspirehep/modules/theme/views.py]: Nit: the sphinx way to do it is like that https://github.com/inspirehep/inspire-next/blob/75d6bcdf28e1c0cb5e4e773a7af4f2e6157b1831/tests/ui/conftest.py#L88

File: inspirehep/modules/records/serializers/schemas/json.py [hunk 4]
```diff
@@ -36,3 +41,98 @@ class RecordSchemaJSONUIV1(Schema):
...
+
+
+class MetadataReferencesSchemaItemV1(Schema):
+    arxiv_eprints = fields.List(fields.Dict())
+    authors = fields.Method('get_authors')
+    collaborations = fields.List(fields.Dict())
+    control_number = fields.Int()
+    dois = fields.List(fields.Dict())
+    publication_info = fields.List(fields.Dict())
+    titles = fields.List(fields.Dict())
```
Comment (michamos) [inline inspirehep/modules/records/serializers/schemas/json.py]: why do you need these fields in references? don't you just need the references?

**Comment by drjova** [comment]
@michamos updated

File: inspirehep/modules/records/serializers/schemas/json.py [hunk 5]
```diff
@@ -36,3 +40,105 @@ class RecordSchemaJSONUIV1(Schema):
...
+                reference.get('arxiv_eprint')),
+            'collaborations': self.prepare_collaborations(
+                reference.get('collaborations', [])),
+            'dois': self.prepare_dois(reference.get('dois', [])),
+            'titles': self.prepare_titles(reference.get('title', [])),
+        })
+        reference = strip_empty_values(reference)
+        return self.only_meaningful_data(reference)
+
+    def only_meaningful_data(self, reference):
```
Comment (michamos) [inline inspirehep/modules/records/serializers/schemas/json.py]: * as discussed IRL, I think it's strange if you don't silently skip references. We probably need to design more generally how to show unlinked references, but that can be done later.
* this and all next methods can be decorated with `@staticmethod`.

File: tests/integration/records/test_records_utils.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+def test_empty_references(isolated_app):
+    results = get_resolved_references([])
+    assert results == []
+
+
+def test_not_existing_references(isolated_app):
+    TestRecordMetadata.create_from_file(
+        __name__, '29177.json', index_name='records-hep')
+    references = [9999999, ]
```
[the remark this answers is not in the document]
Comment (drjova) [reply tests/integration/records/test_records_utils.py]: yep is just styling

File: tests/unit/records/test_references_schemas.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+            ]
+        }
+    }
+
+    result = json.loads(schema.dumps(record).data)
+    assert expected == result
+
+
+@patch('inspirehep.modules.records.serializers.schemas.json.get_resolved_references')
+def test_references_schema_wit_record(mock_get_resolved_references):
```
[the remark this answers is not in the document]
Comment (drjova) [reply tests/unit/records/test_references_schemas.py]: why? :)

**Comment by drjova** [comment]
ci merge please

**Comment by drjova** [comment]
ci test please

**Comment by drjova** [comment]
ci merge please

**Comment by drjova** [comment]
ci merge please

**Comment by drjova** [comment]
ci merge please

**Comment by david-caro** [comment]
It looks like it's failing to start the worker node, because the rt lib is failing to login:
```
2018-06-10T16:06:29.708176698Z [2018-06-10 16:06:29,707] ERROR in app: Failed to initialize entry point: inspire_utils = inspirehep.utils.ext:INSPIREUtils
2018-06-10T16:06:29.708418515Z Traceback (most recent call last):
2018-06-10T16:06:29.708446223Z   File "/virtualenv/bin/celery", line 11, in <module>
2018-06-10T16:06:29.708450628Z     sys.exit(main())
2018-06-10T16:06:29.708487878Z   File "/virtualenv/lib/python2.7/site-packages/celery/__main__.py", line 16, in main
2018-06-10T16:06:29.708513009Z     _main()
2018-06-10T16:06:29.708523628Z   File "/virtualenv/lib/python2.7/site-packages/celery/bin/celery.py", line 322, in main
2018-06-10T16:06:29.708654398Z     cmd.execute_from_commandline(argv)
2018-06-10T16:06:29.708661528Z   File "/virtualenv/lib/python2.7/site-packages/celery/bin/celery.py", line 484, in execute_from_commandline
2018-06-10T16:06:29.70880111Z     super(CeleryCommand, self).execute_from_commandline(argv)))
2018-06-10T16:06:29.70882238Z   File "/virtualenv/lib/python2.7/site-packages/celery/bin/base.py", line 273, in execute_from_commandline
2018-06-10T16:06:29.708966747Z     argv = self.setup_app_from_commandline(argv)
2018-06-10T16:06:29.708978941Z   File "/virtualenv/lib/python2.7/site-packages/celery/bin/base.py", line 475, in setup_app_from_commandline
2018-06-10T16:06:29.709087666Z     self.app = self.find_app(app)
2018-06-10T16:06:29.709099167Z   File "/virtualenv/lib/python2.7/site-packages/celery/bin/base.py", line 497, in find_app
2018-06-10T16:06:29.709228048Z     return find_app(app, symbol_by_name=self.symbol_by_name)
2018-06-10T16:06:29.709254393Z   File "/virtualenv/lib/python2.7/site-packages/celery/app/utils.py", line 355, in find_app
2018-06-10T16:06:29.709376193Z     sym = imp(app)
2018-06-10T16:06:29.709399669Z   File "/virtualenv/lib/python2.7/site-packages/celery/utils/imports.py", line 104, in import_from_cwd
2018-06-10T16:06:29.709478227Z     return imp(module, package=package)
2018-06-10T16:06:29.709498898Z   File "/usr/lib64/python2.7/importlib/__init__.py", line 37, in import_module
2018-06-10T16:06:29.709563766Z     __import__(name)
2018-06-10T16:06:29.709593581Z   File "/code/inspirehep/celery_tests.py", line 63, in <module>
2018-06-10T16:06:29.709629254Z     HEP_ONTOLOGY_FILE=HIGGS_ONTOLOGY_FILE.name,
2018-06-10T16:06:29.70965967Z   File "/virtualenv/lib/python2.7/site-packages/invenio_base/app.py", line 108, in _create_app
2018-06-10T16:06:29.709764434Z     modules=extensions,
2018-06-10T16:06:29.709790658Z   File "/virtualenv/lib/python2.7/site-packages/invenio_base/app.py", line 169, in app_loader
2018-06-10T16:06:29.709860044Z     modules=modules)
2018-06-10T16:06:29.70988754Z   File "/virtualenv/lib/python2.7/site-packages/invenio_base/app.py", line 228, in _loader
2018-06-10T16:06:29.709997107Z     init_func(ep.load())
2018-06-10T16:06:29.710009832Z   File "/virtualenv/lib/python2.7/site-packages/invenio_base/app.py", line 168, in <lambda>
2018-06-10T16:06:29.710069825Z     _loader(app, lambda ext: ext(app), entry_points=entry_points,
2018-06-10T16:06:29.710089648Z   File "/code/inspirehep/utils/ext.py", line 38, in __init__
2018-06-10T16:06:29.710142773Z     self.init_app(app)
2018-06-10T16:06:29.710155683Z   File "/code/inspirehep/utils/ext.py", line 42, in init_app
2018-06-10T16:06:29.71016711Z     self.rt_instance = self.create_rt_instance(app)
2018-06-10T16:06:29.710192328Z   File "/code/inspirehep/utils/ext.py", line 62, in create_rt_instance
2018-06-10T16:06:29.710249109Z     loggedin = tracker.login()
2018-06-10T16:06:29.71025405Z   File "/virtualenv/src/rt/rt.py", line 348, in login
2018-06-10T16:06:29.71042233Z     without_login=True)) == 200
2018-06-10T16:06:29.710432306Z   File "/virtualenv/src/rt/rt.py", line 253, in __request
2018-06-10T16:06:29.710532926Z     raise UnexpectedResponse('Received status code {:d} instead of 200.'.format(response.status_code))
2018-06-10T16:06:29.710556175Z rt.UnexpectedResponse: Received status code 501 instead of 200.
```
The proxy should be allowing the authentication though, that's the `default` scenario there in the e2e tests. Something strange is going on...

**Comment by drjova** [comment]
but it's only related to the proxy, right?

**Comment by david-caro** [comment]
Can't say without checking better. That was working on the previous runs, maybe something changed.

--
David Caro <david.caro@cern.ch>

CERN - RCS-SIS
inSPIRE-HEP High Energy Physics information system
http://inspirehep.net

On Jun 10, 2018 9:33 PM, Harris Tzovanakis <notifications@github.com> wrote:

but it's only related to the proxy, right?

**Comment by drjova** [comment]
ci merge please
